### PR TITLE
BidRequest serialization partially done

### PR DESCRIPTION
**Pull request by emanuele-pirro**

### CODE REVIEW

File: openrtb-json/src/main/scala/com/powerspace/openrtb/json/BidRequestSerde.scala [hunk 1]
```diff
@@ -1,17 +1,114 @@
...
   */
 object BidRequestSerde {
+  import io.circe._
+  import io.circe.syntax._
+  import io.circe.generic.extras.semiauto._
+
+  import JsonUtils._
+  implicit val customConfig: Configuration = Configuration.default.withSnakeCaseMemberNames
+
+  implicit val bannerAdTypeEncoder: Encoder[BannerAdType] = protobufEnumEncoder[BannerAdType]
```
Comment (valdo404) [inline openrtb-json/src/main/scala/com/powerspace/openrtb/json/BidRequestSerde.scala]: I think it is better to use absolute path (ie: `Imp.banner.banneradtype`)

File: openrtb-json/src/main/scala/com/powerspace/openrtb/json/BidRequestSerde.scala [hunk 2]
```diff
@@ -1,17 +1,114 @@
...
+  implicit val audioEncoder: Encoder[Audio] = deriveEncoder[Audio].recodeBooleans.clean
+  implicit val pmpEncoder: Encoder[Pmp] = deriveEncoder[Pmp].recodeBooleans.clean
+  implicit val dealEncoder: Encoder[Deal] = deriveEncoder[Deal].recodeBooleans.clean
+
+  /**
+    * Encoder for the OpenRTB native object.
+    */
+  implicit val nativeEncoder: Encoder[Native] =
+    NoNullsEncoder(native => Seq(
+      //("request", native.requestOneof.value.asJson), //@todo: to be verified
```
Comment (valdo404) [inline openrtb-json/src/main/scala/com/powerspace/openrtb/json/BidRequestSerde.scala]: You may have to provide the case when request is a string.       

```
message Native {
      oneof request_oneof {
        // Request payload complying with the Native Ad Specification.
        // Exactly one of {request, request_native} should be used;
        // this is the OpenRTB-compliant field for JSON serialization.
        string request = 1;

        // Request payload complying with the Native Ad Specification.
        // Exactly one of {request, request_native} should be used;
        // this is an alternate field preferred for Protobuf serialization.
        NativeRequest request_native = 50;
      }
``` 

I think you can use a customized encoder and that the rest of encoder can be derived.

File: openrtb-json/src/main/scala/com/powerspace/openrtb/json/BidRequestSerde.scala [hunk 3]
```diff
@@ -1,17 +1,114 @@
...
+      ("ver", native.ver.asJson),
+      ("api", native.api.map(_.value).asJson),
+      ("battr", native.battr.map(_.value).asJson)
+    ))
+
+  implicit val metricEncoder: Encoder[Metric] = deriveEncoder[Metric].clean()
+
+  implicit val impEncoder: Encoder[Imp] = NoNullsEncoder(imp => {
```
Comment (valdo404) [inline openrtb-json/src/main/scala/com/powerspace/openrtb/json/BidRequestSerde.scala]: derive* ;)

File: openrtb-json/src/main/scala/com/powerspace/openrtb/json/BidResponseSerde.scala [hunk 4]
```diff
@@ -1,13 +1,22 @@
...
 import io.circe.{Decoder, Encoder, Json}
+import scalapb.UnknownFieldSet
 
 /**
   * Serialize and Deserialize an OpenRTB BidResponse
   */
 object BidResponseSerde {
 
+  /**
+    * @todo following three lines are unnecessary and are about full automatic derivation
```
Comment (valdo404) [inline openrtb-json/src/main/scala/com/powerspace/openrtb/json/BidResponseSerde.scala]: We will use it for decoder derivation I guess

File: openrtb-json/src/main/scala/com/powerspace/openrtb/json/BidResponseSerde.scala [hunk 5]
```diff
@@ -1,13 +1,22 @@
...
+    */
+  implicit val configuration: Configuration = Configuration.default.withDefaults
+  implicit val noBidReasonDecoder: Decoder[Option[NoBidReason]] = Decoder.decodeOption[Int].map(_.map(NoBidReason.fromValue))
```
Comment (valdo404) [inline openrtb-json/src/main/scala/com/powerspace/openrtb/json/BidResponseSerde.scala]: `protobufDecoder[NoBidReason]` ;)

File: openrtb-json/src/main/scala/com/powerspace/openrtb/json/util/JsonUtils.scala [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      override def apply(a: A): Json = Json.fromFields(f(a))
+    }.clean(toKeep)
+  }
+
+
+  def protobufEnumEncoder[T <: _root_.scalapb.GeneratedEnum]: Encoder[T] = {
+    (a: T) => Json.fromInt(a.value)
+  }
+
+  implicit val unknownFieldsEncoder: Encoder[UnknownFieldSet] = (_: UnknownFieldSet) => Json.Null
```
Comment (valdo404) [inline openrtb-json/src/main/scala/com/powerspace/openrtb/json/util/JsonUtils.scala]: I think we can define an approach for the unknownfieldset Encoder, by defining a list of PB extensions we want to encode

File: openrtb-json/src/test/scala/com/powerspace/openrtb/json/OpenRtbSerdeTest.scala [hunk 7]
```diff
@@ -56,13 +63,118 @@ class OpenRtbSerdeTest extends FunSuite with GivenWhenThen {
...
+    assert(bannerCursor.downField("w").as[Int].right.get == 10)
+    assert(bannerCursor.downField("api").as[Seq[Int]].right.get == Seq(3))
+    assert(bannerCursor.downField("format").downArray.downField("w").as[Int].right.get == 10)
+
+    val metricCursor = impCursor.downField("metric")
+    assert(metricCursor.downArray.downField("type").as[String].right.get == "type-1")
+
+  }
+
+  private def getBidRequest: BidRequest = {
```
Comment (valdo404) [inline openrtb-json/src/test/scala/com/powerspace/openrtb/json/OpenRtbSerdeTest.scala]: Can you create a `BidRequestFixtures` object ?

File: openrtb-json/src/test/scala/com/powerspace/openrtb/json/OpenRtbSerdeTest.scala [hunk 8]
```diff
@@ -56,13 +63,118 @@ class OpenRtbSerdeTest extends FunSuite with GivenWhenThen {
...
 
     Then("It should return a proper bid request in JSON format")
-    //println(json)
+    println(json)
+    val reqCursor = json.hcursor
+    assert(reqCursor.downField("id").as[String].right.get == "fmySKZNcTFcTPOurFYivufGxMtuSYpen")
+    assert(reqCursor.downField("at").as[Int].right.get == 2)
+    assert(reqCursor.downField("cur").as[Seq[String]].right.get == Seq("EUR"))
+
+    assert(reqCursor.downField("source").downField("pchain").as[String].right.get == "pchain-1")
```
Comment (valdo404) [inline openrtb-json/src/test/scala/com/powerspace/openrtb/json/OpenRtbSerdeTest.scala]: You can have several `Then` clauses

**Review by valdo404 — COMMENTED** [review]
Good stuff

**Comment by valdo404** [comment]
I think you will need to provide serialization for the BidSwitch extensions. This is a truism I guess ;)

File: openrtb-json/src/main/scala/com/powerspace/openrtb/json/BidRequestSerde.scala (hunk 1 above)
Comment (emanuele-pirro) [reply openrtb-json/src/main/scala/com/powerspace/openrtb/json/BidRequestSerde.scala]: These guys are enums and they're all at `openrtb` package level

**Pull request merged**